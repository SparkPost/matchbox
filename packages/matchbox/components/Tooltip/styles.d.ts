export function visibility(props: any): string;
