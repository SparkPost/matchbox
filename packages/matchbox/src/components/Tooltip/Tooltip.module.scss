@import '../../styles/config';
$arrow-margin: 10px;

.Activator {
  position: relative;
  display: inline-block;
}

.hover {
  .Tooltip {
    opacity: 1;
    transform: translate(0, 0) scale(1);
  }
}

.dark {
  .Tip, .Tooltip {
    background: color(gray, 1);
    border: none !important;
  }

  .Content {
    color: color(gray, 10);
  }
}

.top {
  .Tip {
    top: auto;
    bottom: rem(-14);
    margin: 0 0 rem(7);
    border-top: none;
    border-left: none;
    border-bottom-right-radius: border-radius(large);
    border-top: none;
    border-left: none;
    border-bottom: 1px solid color(gray, 8);
    border-right: 1px solid color(gray, 8);
  }

  .Tooltip {
    top: auto;
    bottom: 100%;
    margin: 0 0 $arrow-margin;
  }
}

.left {
  .Tooltip {
    left: auto;
    right: 0;
  }

  .Tip {
    left: auto;
    right: rem(21);
  }
}

.Tip {
  display: block;
  position: absolute;
  top: rem(-14);
  left: rem(21);
  width: rem(16);
  height: rem(16);

  margin: rem(7) 0 0;

  background: color(gray, 10);
  border-top-left-radius: border-radius(large);
  border-top: 1px solid color(gray, 8);
  border-left: 1px solid color(gray, 8);
  transform: rotate(45deg);
  z-index: -1;
}

.Tooltip {
  display: block;
  position: absolute;
  z-index: 1;
  top: 100%;
  left: 0;
  max-height: rem(198);

  background: color(gray, 10);
  margin: $arrow-margin 0 0;
  padding: rem(12);
  box-shadow: shadow(deeper);
  border-radius: border-radius(large);
  border: 1px solid color(gray, 8);

  opacity: 0;
  pointer-events: none;
  transform: translate(0, 0) scale(0.96);
  transition: 0.06s ease-out;
}

.Content {
  position: relative;
  z-index: 1;
  border-radius: border-radius(large);
  max-height: rem(189);
  overflow: hidden;

  word-break: normal;
  text-align: left;
  font-size: font-size(300);
  font-weight: 400;
  line-height: rem(21);
}
