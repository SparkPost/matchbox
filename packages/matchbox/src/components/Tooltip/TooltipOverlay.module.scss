@import '../../styles/config';

.TooltipOverlay {
  position: absolute;
  pointer-events: none;
  z-index: z-index(overlay);
}
