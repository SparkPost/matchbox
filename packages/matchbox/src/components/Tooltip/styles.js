import { tokens } from '@sparkpost/design-tokens';

export const visibility = props => `
  pointer-events: none;
  opacity: ${props.visible ? '1' : '0'};
  visibility: ${props.visible ? 'visible' : 'hidden'};
  transform: scale(${props.visible ? '1' : '0.96'});
  transition:  ${tokens.motionDuration_fast} ${
  props.visible ? tokens.motionEase_out : tokens.motionEase_in
};
`;
