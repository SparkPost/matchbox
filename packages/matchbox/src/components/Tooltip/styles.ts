import { tokens } from '@sparkpost/design-tokens';

export const visibility = (props): string => {
  const { $visible, $state } = props;
  return `
      pointer-events: none;
      opacity: ${$visible && $state === 'entered' ? '1' : '0'};
      visibility: ${$visible && $state === 'entered' ? 'visible' : 'hidden'};
      transform: scale(${$visible && $state === 'entered' ? '1' : '0.96'});
      transition:  ${tokens.motionDuration_fast} ${
    $visible && $state === 'entered' ? tokens.motionEase_out : tokens.motionEase_in
  };`;
};
