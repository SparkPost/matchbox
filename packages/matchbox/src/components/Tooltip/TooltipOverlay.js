import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { Portal } from '../Portal';
import { WindowEvent } from '../WindowEvent';
import { debounce } from '../../helpers/event';
import { getPositionFor, getPreferredDirectionFor } from '../../helpers/geometry';
import styles from './TooltipOverlay.module.scss';

class TooltipOverlay extends Component {
  static displayName = 'TooltipOverlay';

  static defaultProps = {
    eventDebounce: 400
  }

  static propTypes = {
    renderActivator: PropTypes.func.isRequired,
    renderTooltip: PropTypes.func.isRequired,
    eventDebounce: PropTypes.number
  }

  state = {
    position: {
      top: 0,
      left: 0,
      width: 0,
      height: 0
    },
    preferredDirection: {
      top: null,
      bottom: null,
      left: null,
      right: null
    }
  }

  componentDidMount() {
    this.handleMeasurement();
  }

  componentWillReceiveProps() {
    this.handleMeasurement();
  }

  handleMeasurement = () => {
    this.setState({
      preferredDirection: getPreferredDirectionFor(this.activator),
      position: getPositionFor(this.activator)
    });
  }

  render() {
    const { renderTooltip, renderActivator, eventDebounce, portalId } = this.props;
    const { position, preferredDirection } = this.state;

    const tooltipProps = { preferredDirection };
    const activatorProps = {
      activatorRef: (node) => this.activator = node
    };

    return (
      <Fragment>
        <WindowEvent event='resize' handler={debounce(this.handleMeasurement, eventDebounce)} />
        <WindowEvent event='scroll' handler={debounce(this.handleMeasurement, eventDebounce)} />
        {renderActivator(activatorProps)}
        <Portal containerId={portalId}>
          <div className={styles.TooltipOverlay} style={position}>
            {renderTooltip(tooltipProps)}
          </div>
        </Portal>
      </Fragment>
    );
  }
}

export default TooltipOverlay;
