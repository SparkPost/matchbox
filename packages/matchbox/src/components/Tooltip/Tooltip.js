import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import TooltipOverlay from './TooltipOverlay';

import styles from './Tooltip.module.scss';

class Tooltip extends Component {
  static displayName = 'Tooltip';

  static propTypes = {
    content: PropTypes.node,
    /**
     * Disables hover events
     */
    disabled: PropTypes.bool,
    dark: PropTypes.bool,
    active: PropTypes.bool,
    left: PropTypes.bool,
    right: PropTypes.bool,
    top: PropTypes.bool,
    bottom: PropTypes.bool,
    horizontalOffset: PropTypes.string,
    /**
     * Disables automatic positioning
     */
    forcePosition: PropTypes.bool,
    children: PropTypes.node,
    /**
     * These props are provided automatically through the Overlay component
     */
    preferredDirection: PropTypes.shape({
      top: PropTypes.bool,
      bottom: PropTypes.bool,
      left: PropTypes.bool,
      right: PropTypes.bool
    }),
    /**
     * Element ID for the portal that will house tooltips. Appends to body if not provided.
     */
    portalId: PropTypes.string
  };

  static defaultProps = {
    right: true,
    bottom: true,
    horizontalOffset: '0px',
    forcePosition: false,
    preferredDirection: {
      bottom: true,
      left: false,
      right: true,
      top: false
    }
  }

  state = {
    hover: false
  }

  handleMouseOver = () => {
    this.setState({ hover: true });
  }

  handleMouseOut = () => {
    this.setState({ hover: false });
  }

  renderTooltip = ({ preferredDirection }) => {
    const {
      content,
      dark,
      top,
      left,
      horizontalOffset,
      forcePosition,
      disabled,
      width = '210px'
    } = this.props;

    const positionTop = preferredDirection.top || (top && !forcePosition);
    const positionLeft = preferredDirection.left || (left && !forcePosition);

    const wrapperClasses = classnames(
      !disabled && this.state.hover && styles.hover,
      dark && styles.dark,
      positionTop && styles.top,
      positionLeft && styles.left,
    );

    const offset = positionLeft ? { right: horizontalOffset } : { left: horizontalOffset };

    return (
      <span className={wrapperClasses}>
        <span className={styles.Tooltip} style={{ ...offset, width }}>
          <span className={styles.Tip} />
          <div className={styles.Content}>{content}</div>
        </span>
      </span>
    );
  }

  renderActivator = ({ activatorRef }) => (
    <span
      className={styles.Activator}
      onMouseOver={this.handleMouseOver}
      onMouseOut={this.handleMouseOut}
      ref={activatorRef}>
      {this.props.children}
    </span>
  )

  render() {
    return (
      <TooltipOverlay
        eventDebounce={this.props.eventDebounce}
        portalId={this.props.portalId}
        renderTooltip={this.renderTooltip}
        renderActivator={this.renderActivator} />
    );
  }
}

export default Tooltip;
