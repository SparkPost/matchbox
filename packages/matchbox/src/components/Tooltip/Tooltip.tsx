import React from 'react';
import { Transition } from 'react-transition-group';
import styled from 'styled-components';
import {
  border,
  BorderProps,
  color,
  ColorProps,
  compose,
  layout,
  LayoutProps,
  padding,
  PaddingProps,
  typography,
  TypographyProps,
  shadow,
  ShadowProps,
} from 'styled-system';
import { pick } from '@styled-system/props';
import { tokens } from '@sparkpost/design-tokens';
import TooltipOverlay from './TooltipOverlay';
import { Box } from '../Box';
import { ScreenReaderOnly } from '../ScreenReaderOnly';
import { visibility } from './styles';
import { secondsToMS } from '../../helpers/string';

const system = compose(border, color, layout, padding, typography, shadow);

const StyledTooltipContainer = styled(Box)<{
  $visible?: boolean;
  $state?: string;
}>`
  ${visibility}
`;

const StyledContent = styled('div')`
  ${system}
`;

type TooltipProps = {
  as?: 'span' | 'div';
  id: string;
  content?: React.ReactNode;
  /**
   * Disables hover events
   */
  disabled?: boolean;
  /**
   * @deprecated Use system props to set styles
   */
  dark?: boolean;
  left?: boolean;
  right?: boolean;
  top?: boolean;
  bottom?: boolean;
  /**
   * Disables automatic positioning
   */
  forcePosition?: boolean;
  children?: React.ReactNode;
  /**
   * Element ID for the portal that will house tooltips. Appends to body if not provided.
   */
  portalId?: string;
} & BorderProps &
  ColorProps &
  LayoutProps &
  PaddingProps &
  TypographyProps &
  ShadowProps;

function Tooltip(props: TooltipProps): JSX.Element {
  const [show, setshow] = React.useState(false);

  function handleShow() {
    setshow(true);
  }

  function handleHide() {
    setshow(false);
  }

  function renderA11yContent() {
    return (
      <ScreenReaderOnly>
        <span role="tooltip" aria-hidden={!show} id={props.id}>
          {props.content}
        </span>
      </ScreenReaderOnly>
    );
  }

  function renderTooltip({ preferredDirection }) {
    const {
      content,
      dark, // TODO deprecate in favor of system props
      top,
      left,
      // These two props are broken
      right, // Omitting as this is a valid system prop
      bottom, // Omitting as this is a valid system prop
      forcePosition = false,
      disabled,
      width = '13rem',
      ...rest
    } = props;

    const transitionRef = React.useRef();
    const systemProps = pick(rest);
    const positionTop = preferredDirection.top || (top && !forcePosition);
    const positionLeft = preferredDirection.left || (left && !forcePosition);

    const contentStyles = {
      ...systemProps,
      ...(dark ? { bg: 'gray.1000' } : {}), // TODO deprecate in favor of system props
    };

    return (
      <Transition
        mountOnEnter
        unmountOnExit
        in={show}
        timeout={{
          enter: secondsToMS(tokens.motionDuration_medium),
          exit: 0,
        }}
        nodeRef={transitionRef}
      >
        {(state) => (
          <StyledTooltipContainer
            display="block"
            position="absolute"
            top={positionTop ? 'auto' : '100%'}
            bottom={positionTop ? '100%' : 'auto'}
            left={positionLeft ? 'auto' : '0'}
            right={positionLeft ? '0' : 'auto'}
            // Tooltip gap to activator
            mt={positionTop ? '0' : '100'}
            mb={positionTop ? '100' : '0'}
            $visible={!disabled && show}
            $state={state}
            ref={transitionRef}
          >
            <StyledContent
              position="relative"
              fontSize="100"
              lineHeight="100"
              p="200"
              borderRadius="100"
              bg="gray.1000"
              color="white"
              textAlign="left"
              width={width}
              zIndex="default"
              {...contentStyles}
            >
              <Box overflow="hidden" maxHeight="11rem">
                {content}
              </Box>
            </StyledContent>
          </StyledTooltipContainer>
        )}
      </Transition>
    );
  }

  function renderActivator({ activatorRef }) {
    const { as = 'span' } = props;

    return (
      <Box
        as={as}
        // Inline block is required to measure and set height
        display={as === 'span' ? 'inline-block' : null}
        onFocus={handleShow}
        onBlur={handleHide}
        onMouseOver={handleShow}
        onMouseOut={handleHide}
        ref={activatorRef}
      >
        {props.children}
      </Box>
    );
  }

  return (
    <TooltipOverlay
      hideTooltip={handleHide}
      portalId={props.portalId}
      renderA11yContent={renderA11yContent}
      renderTooltip={renderTooltip}
      renderActivator={renderActivator}
      visible={!props.disabled && show}
    />
  );
}

Tooltip.displayName = 'Tooltip';
export default Tooltip;
