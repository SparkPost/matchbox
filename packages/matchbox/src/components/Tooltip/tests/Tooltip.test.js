import React from 'react';
import Tooltip from '../Tooltip';
import { shallow } from 'enzyme';
import cases from 'jest-in-case';

describe('Tooltip', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Tooltip content='content test'>Hola!</Tooltip>);
  });

  const testCases = [
    { name: 'dark', props: { dark: true }},
    { name: 'top left and offset', props: { left: true, top: true, horizontalOffset: '10px' }},
    { name: 'with a width', props: { width: '100px' }}
  ];

  cases('renders tooltip states', (opts) => {
    wrapper.setProps(opts.props);
    expect(wrapper.instance().renderTooltip({ preferredDirection: {
      bottom: true,
      left: false,
      right: true,
      top: false
    }})).toMatchSnapshot();

    expect(wrapper.instance().renderActivator({ activatorRef: jest.fn() })).toMatchSnapshot();
  }, testCases);

  it('should handle mouse events', () => {
    wrapper.instance().handleMouseOver();
    expect(wrapper).toHaveState({ hover: true });
    wrapper.instance().handleMouseOut();
    expect(wrapper).toHaveState({ hover: false });
  });

  it('should render overlay', () => {
    wrapper.setProps({ portalId: 'foo', eventDebounce: 800 });
    expect(wrapper.find('TooltipOverlay').dive()).toMatchSnapshot();
  });
});
