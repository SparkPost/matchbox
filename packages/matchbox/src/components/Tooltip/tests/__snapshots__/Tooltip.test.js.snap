// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders tooltip states dark 1`] = `
<span
  className="dark"
>
  <span
    className="Tooltip"
    style={
      Object {
        "left": "0px",
        "width": "210px",
      }
    }
  >
    <span
      className="Tip"
    />
    <div
      className="Content"
    >
      content test
    </div>
  </span>
</span>
`;

exports[`Tooltip renders tooltip states dark 2`] = `
<span
  className="Activator"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  Hola!
</span>
`;

exports[`Tooltip renders tooltip states top left and offset 1`] = `
<span
  className="top left"
>
  <span
    className="Tooltip"
    style={
      Object {
        "right": "10px",
        "width": "210px",
      }
    }
  >
    <span
      className="Tip"
    />
    <div
      className="Content"
    >
      content test
    </div>
  </span>
</span>
`;

exports[`Tooltip renders tooltip states top left and offset 2`] = `
<span
  className="Activator"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  Hola!
</span>
`;

exports[`Tooltip renders tooltip states with a width 1`] = `
<span
  className=""
>
  <span
    className="Tooltip"
    style={
      Object {
        "left": "0px",
        "width": "100px",
      }
    }
  >
    <span
      className="Tip"
    />
    <div
      className="Content"
    >
      content test
    </div>
  </span>
</span>
`;

exports[`Tooltip renders tooltip states with a width 2`] = `
<span
  className="Activator"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  Hola!
</span>
`;

exports[`Tooltip should render overlay 1`] = `
<Fragment>
  <WindowEvent
    event="resize"
    handler={[Function]}
  />
  <WindowEvent
    event="scroll"
    handler={[Function]}
  />
  <span
    className="Activator"
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  >
    Hola!
  </span>
  <Portal
    containerId="foo"
  >
    <div
      className="TooltipOverlay"
      style={
        Object {
          "height": undefined,
          "left": NaN,
          "top": NaN,
          "width": undefined,
        }
      }
    >
      <span
        className=""
      >
        <span
          className="Tooltip"
          style={
            Object {
              "left": "0px",
              "width": "210px",
            }
          }
        >
          <span
            className="Tip"
          />
          <div
            className="Content"
          >
            content test
          </div>
        </span>
      </span>
    </div>
  </Portal>
</Fragment>
`;
